plasmapy/particles/ionization_state.py

```diff
@@ -2,30 +2,167 @@
 from astropy import units as u
-from numbers import Integral, Real
+
 from plasmapy.particles import Particle, particle_input
-from plasmapy.particles.exceptions import AtomicError, ChargeError, InvalidParticleError
+from plasmapy.particles.exceptions import (
+    ChargeError,
+    InvalidParticleError,
+    ParticleError,
+)
 from plasmapy.utils.decorators import validate_quantities
-from typing import Union, List, Optional
-
 
 _number_density_errmsg = (
     "Number densities must be Quantity objects with units of inverse " "volume."
 )
 
-# TODO: Change `State` into a class with validations for all of the
-# TODO: attributes.
 
-State = collections.namedtuple(
-    "State", ["integer_charge", "ionic_fraction", "ionic_symbol", "number_density"]
-)
+class IonicFraction:
+    """
+    Representation of the ionic fraction for a single ion.
+
+    This class is created when indexing
+    `~plasmapy.particles.IonizationState` or
+    `~plasmapy.particles.IonizationStateCollection`.
+
+    Parameters
+    ----------
+    ion: particle-like
+        The ion for the corresponding ionic fraction.
+
+    ionic_fraction: real number between 0 and 1, optional
+        The fraction of an element or isotope that is at this ionization
+        level.
+
+    number_density: ~astropy.units.Quantity, optional
+        The number density of this ion.
+
+    See Also
+    --------
+    IonizationState
+    IonizationStateCollection
+
+    Examples
+    --------
+    >>> alpha_fraction = IonicFraction("alpha", ionic_fraction=0.31)
+    >>> alpha_fraction.ionic_symbol
+    'He-4 2+'
+    >>> alpha_fraction.integer_charge
+    2
+    >>> alpha_fraction.ionic_fraction
+    0.31
+    """
+
+    def __eq__(self, other):
+
+        try:
+            if self.ionic_symbol != other.ionic_symbol:
+                raise TypeError(
+                    f"Unable to ascertain equality of ionic fractions for "
+                    f"different ions."
+                )
+
+            ionic_fraction_within_tolerance = np.isclose(
+                self.ionic_fraction, other.ionic_fraction, rtol=1e-15,
+            )
+
+            number_density_within_tolerance = u.isclose(
+                self.number_density, other.number_density, rtol=1e-15,
+            )
+
+            return all(
+                [ionic_fraction_within_tolerance, number_density_within_tolerance,]
+            )
+
+        except Exception as exc:
+            raise TypeError(
+                "Unable to ascertain equality between the following objects:\n"
+                f"  {self}\n"
+                f"  {other}"
+            ) from exc
+
+    @particle_input
+    def __init__(self, ion: Particle, ionic_fraction=None, number_density=None):
+        try:
+            self._particle = ion
+            self.ionic_fraction = ionic_fraction
+            self.number_density = number_density
+        except Exception as exc:
+            raise ParticleError("Unable to create IonicFraction object") from exc
```

Comment:
Fun fact! I've been tinkering with dataclasses when scripting recently, and except for the `particle_input` decorator, I think this could have been one - if not for the Py3.7 requirement. Then you wouldn't really have to write a bunch of these utility methods. But of course, that's not a change suggestion! I really wish GitHub would let me differentiate review comments between "change this" and "oh I'm just fawning over f-strings, don't mind me."